### DISCUSSION TITLE
AST Serialization

### DISCUSSION THREAD

**Discussion by ftomassetti** [Strumenta/StarLasu#1] [General]
Right now we have two ways to serialize a ParsingResult or a Node to JSON:
* The "native" JSON serialization. This was created first. It is intended to be as simple as possible
* The EMF based serialization

Having the two of them can cause confusion and duplicated work.

The native one has the advantage to not having dependencies and being simpler to replicate in other platforms. It may be also faster.

The EMF based one can benefit from compatibility to EMF. It can give potential more headaches, but on the other hand EMF is well-proven (even if not all porting of EMF-JSON may be mature).

The question is: should we keep them both or drop one of the two?

**Comment by ftomassetti** [comment]
In addition to this, there is also a native serialization to XML

**Comment by ftomassetti** [comment]
This discussion is created here and not on the Kolasu repository as this is a decision that would make sense to apply on all the *Lasu repositories. The serialization is also used by the multi-platform mechanism, which affects several *Lasu libraries

**Comment by alessiostalla** [comment]
I'm fond of the Ecore based format because they've already solved many problems for us; that said, the EMFJSON format is not standardized, different implementations appear to handle it differently (e.g. EMFJSON vs ecore.js), and it looks like it's abandoned/receiving no interest.

If we decide to bet on our proprietary format though I think we should sit down and design it so it's better optimized (e.g. not writing stuff that is not necessary, not formatting the JSON which is very costly for large files, stuff like that).

**Reply by ftomassetti** [reply]
In general my fear is that EMF is very rich and complex, and some parts evolved over time. For example, the way in which generics are support. It seems that one could either have a type or a generic type, and the whole thing seems not too well thought out. It is also strongly connected to Java. 

It is true that we could simply being more restrictive in the EMF metamodels we generate or use internally, as we do not need to work with arbitrary EMF models. For example, in Kolasu references cannot have multiplicity higher than 1.

The EMF-JSON format has issues with references. It can also be a bit too flexible because of the richness of EMF. For example, when storing a metamodel and a model in the same resource, the model can refer to the EClasses in the metamodel either using relative paths (`/0/myEclass`) or absolute paths (`https://strumenta.com/mypackage/myEclass`). This flexibility means more work for all users of the format such as the multiple *Lasu libraries, and possibly other tools. 

That said, I agree that we could end up recreating something very similar, and we could get things wrong a few times before getting there.

Regarding the formatting JSON, I think it is useful in the rare case in which people want to look into it, but that could be at the very least optional.

**Reply by loradd** [reply]
I think being more restrictive in how the Ecore metamodels should be defined could be a good starting point towards defining our own proprietary format. In this way, we could reuse what we need from Ecore and later change what turns out to be problematic for us.

**Comment by alessiostalla** [comment]
I also think that it could make sense to integrate our own format with Ecore. That is, now we're discussing Ecore + EMFJSON vs no Ecore + our proprietary format, but they can be decoupled. We could still build the Ecore models and metamodels that we're building and use those as the driver of the export, while abandoning EMFJSON in favor of our own format.  

That's because if we drop Ecore altogether we lose not only EMFJSON, but also a somewhat portable metamodel definition, that we're currently using to quickly port an AST from a language to another.

**Reply by ftomassetti** [reply]
One advantage of using our format is that we do not need to go through the conversion to Ecore.
Would our own format be a new serialization format for models containing ASTs? The models would be used by playground or sent by the Kotlin parser to the Pylasu/Kolasu wrappers
Should we instead continue to export the metamodel into XMI or JSON? The metamodels would be used during the generation of the Pylasu/Kolasu classes

**Comment by ftomassetti** [comment]
Other difference: the EMFJSON format may also optionally include in the same file both the metamodel definition and the AST, the custom JSON format cannot at this stage

**Comment by ftomassetti** [comment]
In addition to this we also have XML serialization, which is partially implemented in Kolasu. There is also no deserialization, so I think we should decide if we want to support it properly or remove it in future versions of Kolasu
